### ISSUE TITLE
FFI related problems

### ISSUE THREAD

**Issue by yblee87** [s5z/zsim#52]
Hello,

I faced FFI related problems with the latest version of zsim.
The problem is that zsim enters fast-forwarding mode again after some period of real execution.
This is the console output I got.

```
[H] Starting zsim, built Thu Jul 9 13:46:48 KST 2015 (rev master:69:ac6d5ad:1fc 2+ 2- 5753470a)
[H] Creating global segment, 1024 MBs
[H] Global segment shmid = 458753
[H] Deadlock detection ON
[S 0] Started instance
[S 0] Started RR scheduler, quantum=10000 phases
== Loading device model file '/home/smartcode/Workspace/zsim/ext/DRAMSim2/ini/2Gb_DDR3-1600_x8.ini' ==
== Loading system model file '/home/smartcode/Workspace/zsim/ext/DRAMSim2/system.ini' ==
===== MemorySystem 0 =====
CH. 0 TOTAL_STORAGE : 4096MB | 2 Ranks | 8 Devices per rank
===== MemorySystem 1 =====
CH. 1 TOTAL_STORAGE : 4096MB | 2 Ranks | 8 Devices per rank
[S 0] Hierarchy: [ l1i-0 l1d-0 ] -> l2-0
[S 0] Hierarchy: [ l1i-1 l1d-1 ] -> l2-1
[S 0] Hierarchy: [ l1i-2 l1d-2 ] -> l2-2
[S 0] Hierarchy: [ l1i-3 l1d-3 ] -> l2-3
[S 0] Hierarchy: [ l2-0 l2-1 l2-2 l2-3 ] -> l3-0b0..l3-0b3
[S 0] Initialized system
[S 0] HDF5 backend: Opening /home/smartcode/Workspace/zsim/run/429.mcf/zsim.h5
[S 3] Started instance
[S 1] Started instance
[S 0] HDF5 backend: Created table, 3752 bytes/record, 280 records/write
[S 0] HDF5 backend: Opening /home/smartcode/Workspace/zsim/run/429.mcf/zsim-ev.h5
[S 2] Started instance
[S 0] HDF5 backend: Created table, 3752 bytes/record, 35 records/write
[S 0] HDF5 backend: Opening /home/smartcode/Workspace/zsim/run/429.mcf/zsim-cmp.h5
[S 0] HDF5 backend: Created table, 2336 bytes/record, 1 records/write
[S 0] Initialization complete
[S 0] Started process, PID 27358
[S 1] Started process, PID 27359
[S 1] procMask: 0x400000000000000
[S 0] procMask: 0x0
[S 1] [1] Adjusting clocks, domain 0, de-ffwd 0
[S 1] FFI mode initialized, 2 ffiPoints
[S 0] FFI mode initialized, 2 ffiPoints
[S 2] Started process, PID 27360
[S 2] procMask: 0x800000000000000
[H] Attached to global heap
[S 2] FFI mode initialized, 2 ffiPoints
[S 3] Started process, PID 27361
[S 1] vDSO info initialized
[S 0] vDSO info initialized
[S 3] procMask: 0xc00000000000000
[S 3] FFI mode initialized, 2 ffiPoints
[S 2] vDSO info initialized
[S 3] vDSO info initialized
[S 3] FF control Thread TID 27375
[S 3] FF thread 0 starting
[S 1] FF control Thread TID 27372
[S 0] FF control Thread TID 27373
[S 0] Started contention simulation thread 0
[S 0] Started scheduler watchdog thread
[S 1] FF thread 0 starting
[S 0] FF thread 0 starting
[S 2] FF control Thread TID 27374
[S 2] FF thread 0 starting


MCF SPEC CPU2006 version 1.10
Copyright (c) 1998-2000 Zuse Institut Berlin (ZIB)
Copyright (c) 2000-2002 Andreas Loebel & ZIB
Copyright (c) 2003-2005 Andreas Loebel
MCF SPEC CPU2006 version 1.10

Copyright (c) 1998-2000 Zuse Institut Berlin (ZIB)
Copyright (c) 2000-2002 Andreas Loebel & ZIB
Copyright (c) 2003-2005 Andreas Loebel


MCF SPEC CPU2006 version 1.10
Copyright (c) 1998-2000 Zuse Institut Berlin (ZIB)
Copyright (c) 2000-2002 Andreas Loebel & ZIB
Copyright (c) 2003-2005 Andreas Loebel


MCF SPEC CPU2006 version 1.10
Copyright (c) 1998-2000 Zuse Institut Berlin (ZIB)
Copyright (c) 2000-2002 Andreas Loebel & ZIB
Copyright (c) 2003-2005 Andreas Loebel

nodes                      : 25137
active arcs                : 260767
nodes                      : 25137
active arcs                : 260767
nodes                      : 25137
active arcs                : 260767
nodes                      : 25137
active arcs                : 260767
[S 0] ffiPoint reached, 1000000001 instrs, limit 1000000000
[S 0] FFI: Exiting fast-forward
[S 0] [0] Adjusting clocks, domain 0, de-ffwd 1
[S 0] Thread 0 starting
[S 0] Simulation paused due to synced fast-forwarding
[S 3] ffiPoint reached, 1000000001 instrs, limit 1000000000
[S 3] FFI: Exiting fast-forward
[S 3] [3] Adjusting clocks, domain 0, de-ffwd 1
[S 3] Thread 0 starting
[S 1] ffiPoint reached, 1000000001 instrs, limit 1000000000
[S 1] FFI: Exiting fast-forward
[S 1] [1] Adjusting clocks, domain 0, de-ffwd 1
[S 1] Thread 0 starting
[S 2] ffiPoint reached, 1000000001 instrs, limit 1000000000
[S 2] FFI: Exiting fast-forward
[S 2] [2] Adjusting clocks, domain 0, de-ffwd 1
[S 2] Thread 0 starting
[S 0] Synced fast-forwarding done, resuming simulation
writing vis file to ./results/2Gb_DDR3-1600_x8/4GB.2Ch.2R.scheme7.close_page.64TQ.64CQ.RtB.pRank.vis
DRAMSim2 Clock Frequency =800000000Hz, CPU Clock Frequency=3200000000Hz
[S 0] FFI: Entering fast-forward for process 0
[S 0] Thread 0 entering fast-forward
[S 0] [0] Internal exception detected:
[S 0] [0]  Code: 1
[S 0] [0]  Address: 0x7ffff65cadc1
[S 0] [0]  Description: Exception Code: ACCESS_INVALID_ADDRESS. Exception Address = 0x7ffff65cadc1. Access Type: UNKNOWN. Access Address = 0x000000000
[S 0] [0]  Caused by invalid access to address 0x0
[S 1] Simulation paused due to synced fast-forwarding
[S 0] [0] Backtrace (9/40 max frames)
[S 0] [0]  /home/smartcode/Workspace/zsim/build/opt/zsim.cpp:1397 / InternalExceptionHandler
[S 0] [0]  regstrlcpy.c:0 / LEVEL_PINCLIENT::IEH_CALLBACKS::NotifyInternalException(unsigned int, LEVEL_BASE::EXCEPTION_INFO*, LEVEL_VM::CONTEXT*)
[S 0] [0]  /home/smartcode/Workspace/zsim/ext/pin/pin-2.14-71313-gcc.4.4.7-linux/intel64/bin/pinbin(_ZN8LEVEL_VM12SIGNALS_IMPL19InternalHandlerSyncEiPN7BARECRT8SIGXINFOEPN5PINVM11ISIGCONTEXTEPPKNS_14SCT_ATTRIBUTESEPNS_5PCTXTEPj+0x444) [0x3043a9454]
[S 0] [0]  /home/smartcode/Workspace/zsim/ext/pin/pin-2.14-71313-gcc.4.4.7-linux/intel64/bin/pinbin(_ZN8LEVEL_VM12SIGNALS_IMPL20HandlePhysicalSignalEPN7BARECRT8SIGXINFOEPN5PINVM11ISIGCONTEXTE+0x124) [0x3043aa1f4]
[S 0] [0]  /home/smartcode/Workspace/zsim/ext/pin/pin-2.14-71313-gcc.4.4.7-linux/intel64/bin/pinbin(_ZN5PINVM28SIGNAL_DETAILS_LINUX_INTEL6415InternalHandlerEiPN7BARECRT8SIGXINFOEPv+0xe8) [0x304438c88]
[S 0] [0]  /home/smartcode/Workspace/zsim/ext/pin/pin-2.14-71313-gcc.4.4.7-linux/intel64/bin/pinbin(BARECRT_SigReturnRt+0) [0x30446603c]
[S 0] [0]  /home/smartcode/Workspace/zsim/build/opt/zsim.cpp:515 / TakeBarrier(unsigned int, unsigned int)
[S 0] [0]  /home/smartcode/Workspace/zsim/build/opt/ooo_core.cpp:518 / OOOCore::BblFunc(unsigned int, unsigned long, BblInfo*)
[S 0] [0]  [0x7fffe39bfa5c]
C: Tool (or Pin) caused signal 11 at PC 0x7ffff65cadc1
[H] Child 27358 done
[H] Panic on build/opt/zsim_harness.cpp:123: Child 27358 (idx 0) exit was anomalous, killing simulation
[H] WARN: Hard death at exit (3 children running), killing the whole process tree
Killed
```

Any comment would be appreciated.
Thank you.

**Comment by gaomy3832** [comment]
I am afraid my last bug fix in commit e2f1f7b8922f6defbea68c89821f76a597328267 introduced a new bug. Please try the following patch and let me know. If it works, I will submit another pull request to fix it:
```
--- a/src/zsim.cpp
+++ b/src/zsim.cpp
@@ -509,11 +509,11 @@ uint32_t TakeBarrier(uint32_t tid, uint32_t cid) {
         info("Termination condition met, exiting");
         zinfo->sched->leave(procIdx, tid, newCid);
         SimEnd(); //need to call this on a per-process basis...
+    } else {
+        // Set fPtrs to those of the new core after possible context switch
+        fPtrs[tid] = cores[tid]->GetFuncPtrs();
     }

-    // Set fPtrs to those of the new core after possible context switch
-    fPtrs[tid] = cores[tid]->GetFuncPtrs();
-
     return newCid;
 }
```

**Comment by yblee87** [comment]
Your patch works like a charm :)
Thank you very much.
